cirq/ops/pauli_string.py

```diff
@@ -240,6 +239,189 @@ def _apply_unitary_(self, args: protocols.ApplyUnitaryArgs):
+            return self._expectation_from_wavefunction(state, qubit_map)
+
+        raise ValueError("Input array does not represent a wavefunction with "
+                         "shape `(2 ** n,)` or `(2, 2, ..., 2)`.")
+
+    def _expectation_from_wavefunction(self,
+                                       state: np.ndarray,
+                                       qubit_map: Dict[raw_types.Qid, int]
+                                       ) -> float:
+        """Evaluate the expectation of this PauliString given a wavefunction.
+
+        This method does not provide input validation. See
+        `PauliString.expectation_from_wavefunction` for function description.
+        """
+        if len(state.shape) == 1:
+            num_qubits = state.shape[0].bit_length() - 1
+            state = np.reshape(state, (2,) * num_qubits)
+
+        ket = np.copy(state)
+        for qubit, pauli in self.items():
+            buffer = np.empty(ket.shape, dtype=state.dtype)
+            args = protocols.ApplyUnitaryArgs(target_tensor=ket,
+                                              available_buffer=buffer,
+                                              axes=(qubit_map[qubit],))
+            ket = protocols.apply_unitary(pauli, args)
+
+        return self.coefficient * np.asscalar(
+            np.tensordot(state.conj(), ket, axes=len(ket.shape)))
+
+    def expectation_from_density_matrix(
+        self,
+        state: np.ndarray,
+        qubit_map: Optional[Dict[raw_types.Qid, int]]=None) -> float:
+        r"""Evaluate the expectation of this PauliString given a density matrix.
+
+        Compute the expectation value of this PauliString with respect to an
+        array representing a density matrix. By convention expectation values
+        are defined for Hermitian operators, and so this method will fail if
+        this PauliString is non-Hermitian.
+
+        `state` must be an array representation of a density matrix and have
+        shape `(2 ** n, 2 ** n)` or `(2, 2, ..., 2)` (2*n entries), where
+        `state` is expressed over n qubits.
+
+        # FIXME:
+        If `qubit_map` is not provided, default behavior is to assume that
+        `state` is defined over this PauliString's qubits in ascending order of
+        index. For example if `state` represents $|0\rangle |+\rangle$ and
+        `q0, q1 = cirq.LineQubit.range(2)` then:
+
+            cirq.X(q0).expectation(state) = 0
+            cirq.X(q0).expectation(state, qubit_map={q0: 1, q1: 0}) = 1
+
+        Args:
+            state: An array representing a wavefunction or density matrix.
+            qubit_map: A map from all qubits used in this PauliString to the
+            indices of the qubits that `state` is defined over.
+
+        Returns:
+            The expectation value of the input state.
+
+        Raises:
+            NotImplementedError if this PauliString is non-Hermitian.
+        """
+        if abs(self.coefficient.imag) > 0.0001:
```

Comment:
Most of this validation is common to all four of these expectation methods. It could be extracted into a utility method `_validate_qubit_mapping`.

You could also check that the integers being provided make sense, given the size of the state vector or density matrix.

Reply:
done